src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Remote.cs

```diff
@@ -411,13 +414,300 @@ private IEnumerable<DependencyDetail> CalculateLeavesOfCoherencyTrees(IEnumerabl
+        ///     version details files anyway.
+        /// </remarks>
+        private async Task<List<DependencyUpdate>> GetRequiredStrictCoherencyUpdatesAsync(
+            IEnumerable<DependencyDetail> dependencies,
+            IRemoteFactory remoteFactory)
+        {
+            List<DependencyUpdate> toUpdate = new List<DependencyUpdate>();
+            IEnumerable<DependencyDetail> leavesOfCoherencyTrees = CalculateLeavesOfCoherencyTrees(dependencies);
+
+            if (!leavesOfCoherencyTrees.Any())
+            {
+                // Nothing to do.
+                return toUpdate;
+            }
+
+            List<CoherencyError> coherencyErrors = new List<CoherencyError>();
+
+            // Cache of dependencies. Key is "<repo>@<sha>".
+            Dictionary<string, IEnumerable<DependencyDetail>> dependenciesCache =
+                new Dictionary<string, IEnumerable<DependencyDetail>>();
+            // Cache of builds with assets. Key is "<repo>@<sha>".
+            Dictionary<string, List<Build>> buildCache = new Dictionary<string, List<Build>>();
+            // Cache of nuget config files for further build disambiguation. Key is "<repo>@<sha>".
+            Dictionary<string, IEnumerable<string>> nugetConfigCache = new Dictionary<string, IEnumerable<string>>();
+
+            // Now make a walk over coherent dependencies. Note that coherent dependencies could make
+            // a chain (A->B->C). In all cases we need to walk to the head of the chain, keeping track
+            // of all elements in the chain, then updating then in reverse order (C, B, A).
+            foreach (DependencyDetail dependency in leavesOfCoherencyTrees)
+            {
+                // Build the update stack.
+                // Walk to head of dependency tree, keeping track of elements along the way.
+                // If we hit a pinned dependency in the walk, that means we can't move
+                // the dependency and therefore it is effectively the "head" of the subtree.
+                // We will still visit all the elements in the chain eventually in this algorithm:
+                // Consider A->B(pinned)->C(pinned)->D.
+                Stack<DependencyDetail> updateStack = new Stack<DependencyDetail>();
+                {
+                    DependencyDetail currentDependency = dependency;
+                    while (!string.IsNullOrEmpty(currentDependency.CoherentParentDependencyName) && !currentDependency.Pinned)
+                    {
+                        updateStack.Push(currentDependency);
+                        DependencyDetail parentCoherentDependency = dependencies.FirstOrDefault(d =>
+                            d.Name.Equals(currentDependency.CoherentParentDependencyName, StringComparison.OrdinalIgnoreCase));
+                        currentDependency = parentCoherentDependency ?? throw new DarcException($"Dependency {currentDependency.Name} has non-existent parent " +
```

Comment:
IMO a plain "if" would be more readable here.

Reply:
I dislike the line length, but [`?? throw`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/throw#the-throw-expression) is idiomatic for this kind of quick null check, and IMO nice to make it instantly clear to those familiar with it what's happening here.